### DISCUSSION TITLE
ESP and LAN8720A Ethernet

### DISCUSSION THREAD

**Discussion by DatanoiseTV** [lathoub/Arduino-AppleMIDI-Library#145] [Q&A]
![image](https://user-images.githubusercontent.com/6614616/153597372-a692c9f6-7cea-4e79-a6cc-d07beefd6a86.png)

I have recently made a board (which I am going to release as open source hardware) which has LAN8720A ethernet and an ESP32-WROVER-IE with 16mb flash. I am primarely targetting esp-idf, but also want to support arduino ide / platformio and the applemidi library.

Is there an example for the LAN8720? I cannot get the applemidi library to compile.

```
#define ETH_ADDR 1


#define ETH_RESET 14
#include <ETH.h>
.....

src/main.cpp:9:27: error: 'EthernetUDP' was not declared in this scope
 APPLEMIDI_CREATE_INSTANCE(EthernetUDP, MIDI, "AppleMIDI-Arduino", DEFAULT_CONTROL_PORT);
                           ^
.pio/libdeps/esp32-evb/AppleMIDI/src/AppleMIDI.h:347:43: note: in definition of macro 'APPLEMIDI_CREATE_INSTANCE'
     APPLEMIDI_NAMESPACE::AppleMIDISession<Type> Apple##Name(SessionName, Port); \
```

**Comment by lathoub** [comment]
What mechanism are you using to connect to the LAN8729A? Through SPI?

**Reply by lathoub** [reply]
If you are using SPI, use [EthernetENC](https://github.com/jandrassy/EthernetENC) and read the notes on the use of the ESP32 - good luck

**Comment by DatanoiseTV** [comment]
@lathoub No, the ESP32 has an integrated Ethernet MAC and the LAN8720A PHY is attached via the RGMII interface.

**Reply by lathoub** [reply]
Apart from the wESP32, I have not used the RGMII interface

**Reply by DatanoiseTV** [reply]
Probably we could start looking at https://github.com/khoih-prog/WebServer_WT32_ETH01

**Reply by lathoub** [reply]
OK - the library works fine with the [wESP32](https://wesp32.com) that also uses the RMII interface to access PHY.

Take a look at 
https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/7fbf1e874d037064119006e03d31f73d25bd91c1/src/AppleMIDI.h#L400-L402

somehow the platform identifier don't seem to work and the lib defaults to the default Ethernet lib (hence the use of `EthernetUDP` in your error log)

I suggest to find out what the platform identifier is for your platform (apparently not `ESP32` or `ARDUINO_ARCH_ESP32`), and let me know (through a PR). (Alternatively, alter `AppleMIDI.h` and force the use of 
`    APPLEMIDI_CREATE_INSTANCE(WiFiUDP, MIDI, "AppleMIDI-ESP32", DEFAULT_CONTROL_PORT);
`

**Reply by lathoub** [reply]
@DatanoiseTV did you get it to work?
